Mathlib/Analysis/Convex/Cone/Proper.lean

```diff
@@ -30,77 +32,107 @@ The next steps are:
-  ⟨toPointedCone⟩
+-- TODO: add `ConvexConeClass` that extends `SetLike` and replace the below instance
+instance : SetLike (ProperCone R E) E where
+  coe C := C.carrier
+  coe_injective' _ _ h := ProperCone.toPointedCone_injective <| SetLike.coe_injective h
 
-theorem toPointedCone_injective : Function.Injective ((↑) : ProperCone 𝕜 E → PointedCone 𝕜 E) :=
-  fun S T h => by cases S; cases T; congr
+@[ext] lemma ext (h : ∀ x, x ∈ C₁ ↔ x ∈ C₂) : C₁ = C₂ := SetLike.ext h
 
--- TODO: add `ConvexConeClass` that extends `SetLike` and replace the below instance
-instance : SetLike (ProperCone 𝕜 E) E where
-  coe K := K.carrier
-  coe_injective' _ _ h := ProperCone.toPointedCone_injective (SetLike.coe_injective h)
+@[simp] lemma mem_toPointedCone : x ∈ C.toPointedCone ↔ x ∈ C := .rfl
 
-@[ext]
-theorem ext {S T : ProperCone 𝕜 E} (h : ∀ x, x ∈ S ↔ x ∈ T) : S = T :=
-  SetLike.ext h
+protected lemma pointed_toConvexCone (C : ProperCone R E) : (C : ConvexCone R E).Pointed :=
+  C.toPointedCone.pointed_toConvexCone
 
-@[simp]
-theorem mem_coe {x : E} {K : ProperCone 𝕜 E} : x ∈ (K : PointedCone 𝕜 E) ↔ x ∈ K :=
-  Iff.rfl
+protected lemma nonempty (C : ProperCone R E) : (C : Set E).Nonempty := C.toSubmodule.nonempty
+protected lemma isClosed (C : ProperCone R E) : IsClosed (C : Set E) := C.isClosed'
 
-instance instZero (K : ProperCone 𝕜 E) : Zero K := PointedCone.instZero (K.toSubmodule)
+protected nonrec lemma smul_mem (C : ProperCone R E) (hx : x ∈ C) (hr : 0 ≤ r) : r • x ∈ C :=
+  C.smul_mem ⟨r, hr⟩ hx
+
+section T1Space
+variable [T1Space E]
+
+lemma mem_bot : x ∈ (⊥ : ProperCone R E) ↔ x = 0 := .rfl
+
+@[simp, norm_cast] lemma toPointedCone_bot : (⊥ : ProperCone R E).toPointedCone = ⊥ := rfl
+
+end T1Space
+
+/-- The closure of image of a proper cone under a `ℝ`-linear map is a proper cone. We
+use continuous maps here so that the comap of f is also a map between proper cones. -/
+abbrev comap (f : E →L[R] F) (C : ProperCone R F) : ProperCone R E :=
+  ClosedSubmodule.comap (f.restrictScalars R≥0) C
 
-protected theorem nonempty (K : ProperCone 𝕜 E) : (K : Set E).Nonempty :=
-  ⟨0, by { simp_rw [SetLike.mem_coe, ← ProperCone.mem_coe, Submodule.zero_mem] }⟩
+@[simp] lemma comap_id (C : ProperCone R F) : C.comap (.id _ _) = C := rfl
 
-protected theorem isClosed (K : ProperCone 𝕜 E) : IsClosed (K : Set E) :=
-  K.isClosed'
+@[simp] lemma coe_comap (f : E →L[R] F) (C : ProperCone R F) : (C.comap f : Set E) = f ⁻¹' C := rfl
+
+lemma comap_comap (g : F →L[R] G) (f : E →L[R] F) (C : ProperCone R G) :
+    (C.comap g).comap f = C.comap (g.comp f) := rfl
+
+lemma mem_comap {C : ProperCone R F} {f : E →L[R] F} : x ∈ C.comap f ↔ f x ∈ C := .rfl
+
+variable [ContinuousAdd F] [ContinuousConstSMul R F]
+
+/-- The closure of image of a proper cone under a linear map is a proper cone.
+
+We use continuous maps here to match `ProperCone.comap`. -/
+abbrev map (f : E →L[R] F) (C : ProperCone R E) : ProperCone R F :=
+  ClosedSubmodule.map (f.restrictScalars R≥0) C
```

Comment:
I don't really buy "it matches" as a good enough reason to have the input non-generalised...? 
(But "it's how it was before" is a good enough reason, so this is not an objection :) )

Reply:
Yeah, just moving this around. I am not convinced myself of the design